File: JavaSolutions/HashTable/Easy/FirstUniqueChar.java
/*

387. First Unique Character in a String

https://leetcode.com/problems/first-unique-character-in-a-string/

s = "leetcode"
return 0.

s = "loveleetcode",
return 2.

*/


class FirstUniqChar {
    public int firstUniqChar(String s) {
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        int n = s.length();
        
        for (int i = 0; i< n; i++){
            char c = s.charAt(i);
            map.put(c, map.getOrDefault(c, 0) + 1);
            
            }
        for (int i = 0 ; i < n; i++){
            if (map.get(s.charAt(i)) == 1) 
                return i;            
        }
        return -1;
    }
}

File: Python_Solutions/String/FirstUniqueChar.py
"""
387. First Unique Character in a String
https://leetcode.com/problems/first-unique-character-in-a-string/

Given a string, find the first non-repeating character in it and return it's index. If it doesn't exist, return -1.

Examples:

s = "leetcode"
return 0.

s = "loveleetcode",
return 2.

"""


from collections import Counter

class FirstUniqChar:
    def firstUniqChar(self, s: str) -> int:
        # Counter -> fast
    	for i,j in Counter(s).items(): 
    		if j == 1: return(s.index(i))
    	return -1
    
    def firstUniqChar2(self, s: str) -> int:
        # Dict
        d = {}
        for l in s:
            if l not in d: d[l] = 1
            else: d[l] +=1
        
        index = -1
        for i, char in enumerate(s):
            if d[char] == 1:
                index = i
                break
        return index
        
        

